### ISSUE TITLE
Using epochs in tf.train.slice_input_producer gives uninitialized value when combined with batching

### ISSUE THREAD

**Issue by kalle** [tensorflow/tensorflow#3848]
### Environment info
MacOS X 10.11.4 no CUDA
Tensorflow 0.10.0rc0, installed precompiled pip

### Problem

I have my test data saved as `XXX_input.png` and `XXX_cb.png` (for label) and tried to use the
python code below to create the batched input tensors:

```python
def _load_png_file(path, name):
    file_content = tf.read_file(path)
    png = tf.image.decode_png(file_content, channels=3)
    png = tf.slice(png, [0, 0, 0], [ 128, 128, 3])
    return tf.image.convert_image_dtype(png, dtype=tf.float32, name=name)

def load_data(path):
    if not tf.gfile.Exists(path):
        raise ValueError("Datadir {} don't exists!".format(path))
    photo_files = []                                                                
    cb_files = []
    for file_name in os.listdir(path):                                                  
        if file_name.endswith("_input.png"):
            file_path = os.path.join(path, file_name)                                        
            photo_files.append(file_path)
            cb_files.append(file_path.replace("_input.png", "_cb.png"))

    photos = tf.convert_to_tensor(photo_files, dtype=tf.string)
    cbs = tf.convert_to_tensor(cb_files, dtype=tf.string)
    input_queue = tf.train.slice_input_producer([photos, cbs],          
            num_epochs=FLAGS.num_epochs, # <-- delete this line to make it work
            shuffle=True)
                                                         
    photo = _load_png_file(input_queue[0], "photo")
    cb = _load_png_file(input_queue[1], "cb")
    capacity = FLAGS.min_after_dequeue + 3 * FLAGS.batch_size                       
    photo_batch, cb_batch = tf.train.batch([photo, cb],
            capacity=capacity,                                                             
            batch_size=FLAGS.batch_size)
    return photo_batch, cb_batch
```

When I try to use it it will generate an error as below. When I remove the `num_epochs` argument to `tf.train.slice_input_producer` everything works as expected (net converge etc) so I don't there is something wrong with the rest of the code.

```
E tensorflow/core/client/tensor_c_api.cc:485] Attempting to use uninit[221/1918]
lue input_producer/input_producer/fraction_of_32_full/limit_epochs/epochs
         [[Node: input_producer/input_producer/fraction_of_32_full/limit_epochs/
CountUpTo = CountUpTo[T=DT_INT64, _class=["loc:@input_producer/input_producer/fraction_of_32_full/limit_epochs/epochs"], limit=2, _device="/job:localhost/replic
a:0/task:0/cpu:0"](input_producer/input_producer/fraction_of_32_full/limit_epochs/epochs)]]
ERROR:tensorflow:Exception in QueueRunner: Attempting to use uninitialized value
 input_producer/input_producer/fraction_of_32_full/limit_epochs/epochs
         [[Node: input_producer/input_producer/fraction_of_32_full/limit_epochs/CountUpTo = CountUpTo[T=DT_INT64, _class=["loc:@input_producer/input_producer/fr
action_of_32_full/limit_epochs/epochs"], limit=2, _device="/job:localhost/replic
a:0/task:0/cpu:0"](input_producer/input_producer/fraction_of_32_full/limit_epoch
s/epochs)]]
Caused by op u'input_producer/input_producer/fraction_of_32_full/limit_epochs/Co
untUpTo', defined at:
  File "train.py", line 94, in <module>
    tf.app.run()
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packag
es/tensorflow/python/platform/app.py", line 30, in run
    sys.exit(main(sys.argv))
  File "train.py", line 91, in main
    train()
  File "train.py", line 30, in train
    x, y = cbnet.load_data("/Users/kalle/Development/nn_data/train")
  File "/Users/kalle/Development/tensorflow_cbnet/cbnet.py", line 42, in load_data
    shuffle=True)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packag
es/tensorflow/python/training/input.py", line 266, in slice_input_producer
    shared_name=shared_name)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packages/tensorflow/python/training/input.py", line 223, in range_input_producer
    shared_name, name, "fraction_of_%d_full" % capacity)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packag
es/tensorflow/python/training/input.py", line 133, in input_producer
    input_tensor = limit_epochs(input_tensor, num_epochs)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packages/tensorflow/python/training/input.py", line 84, in limit_epochs
    counter = epochs.count_up_to(num_epochs)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packag
es/tensorflow/python/ops/variables.py", line 577, in count_up_to
    return state_ops.count_up_to(self._variable, limit=limit)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packages/tensorflow/python/ops/gen_state_ops.py", line 127, in count_up_to
    result = _op_def_lib.apply_op("CountUpTo", ref=ref, limit=limit, name=name)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packag
es/tensorflow/python/framework/op_def_library.py", line 703, in apply_op
    op_def=op_def)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2310, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/Users/kalle/Development/tensorflow_cbnet/.env/lib/python2.7/site-packag
es/tensorflow/python/framework/ops.py", line 1232, in __init__
    self._traceback = _extract_stack()
```

**Comment by fwalch** [comment]
You're probably missing a `session.run(tf.initialize_local_variables())`.

**Comment by kalle** [comment]
No, I'm doing that. As I said, I can train the net if I remove the `num_epochs` argument to `tf.train.slice_input_producer` and handle the epochs "manually" in the train loop.

**Comment by yaroslavvb** [comment]
The error says that "input_producer/input_producer/
fraction_of_32_full/limit_epochs/epochs" has not been initialized, which
should be initialized by running "tf.initialize_local_variables()"

Can you check that this variable is present the local variables collection (
[v.name for v in tf.get_collection(tf.GraphKeys.LOCAL_VARIABLES)]) or the
regular collection ([v.name for v in
tf.get_collection(tf.GraphKeys.LOCAL_VARIABLES)])

**Comment by fwalch** [comment]
@kalle Note that `initialize_*local*_variables` is not included in `initialize_all_variables`. If you remove the `num_epochs` argument, your graph shouldn't have any local variables, which is why it works in this case.

**Comment by kalle** [comment]
Aha, @fwalch was right. I thought all variables were initialized when I ran `initialize_all_variables()` so I missed the local initialization. My bad.

**Comment by lucasdavid** [comment]
> Just curious. Under what circumstance do you not want to initialize the local variables?

I'm not sure, but maybe when the training was initiated, but interrupted for some reason; and now you want to restart from the point you left instead of going back to epoch 1..?

**Comment by kalle** [comment]
@lucasdavid Thanks for chiming in. But if you want to continue the training you wouldn't run `initialize_all_variables()` either?  :-)

I think TensorFlow should have three functions:

* `initialize_all_variables()` that really took care of *all* variables.
* `initialize_global_variables()` and `initialize_local_variables()` that did the more specific initializations.

If you are doing something clever/unusual that only require initialization of one kind of variable you are probably pretty skilled with TF and would know there are two different kind of variables.

**Comment by lucasdavid** [comment]
Okay, maybe my example wasn't really a good. I agree that a function called `initialize_all_variables` which does not initialize all variables seems counter intuitive, but there's probably a reason for that. 

Also, following your organization (in which `all_variables` means all variables), you could easily achieve any initialization without the "global_variables" collection:
```py
    with tf.Session() as s:
        # All of them.
        s.run(tf.initialize_all_variables())

        # Local ones.
        s.run(tf.initialize_local_variables())

        # "Global" ones.
        global_variables = set(tf.all_variables()) -set(tf.local_variables())
        s.run(tf.initialize_variables(global_variables))
```

**Comment by jrajagopal** [comment]
@lucasdavid, I ran into the same issue as @kalle, where setting num_epochs gave me the "Attempting to use un-initialized value" error. And like @kalle, I was initializing all variables like I am supposed to. 

To fix it, I tried running tf.initialize_local_variables, but with num_epochs set, I still get the same error. I even tried including the global variable initialization, if only to cover my bases, and that didn't help either.

I can control the number of epochs through a For loop and catch the exception using Try-Except when the loop ends. But it would be more elegant to not have to catch that exception in the first place.

**Comment by rgobbel** [comment]
+1 Seeing the same problem with a very simple test program:

```
from __future__ import print_function

import tensorflow as tf
import numpy as np

N_EXAMPLES = 100
BATCH_SIZE = 4
N_EPOCHS = 3

steps_per_epoch = N_EXAMPLES / BATCH_SIZE

with tf.Graph().as_default():
    init_op = tf.group(tf.initialize_all_variables(), tf.initialize_local_variables())

    a = tf.convert_to_tensor(np.arange(N_EXAMPLES))
    b = tf.convert_to_tensor(np.arange(N_EXAMPLES))

    aa, bb = tf.train.slice_input_producer([a, b], shuffle=True, seed=1, num_epochs=N_EPOCHS)
    batch = tf.train.batch([aa, bb], batch_size=BATCH_SIZE)

    s = tf.Session()

    s.run(init_op)

    coord = tf.train.Coordinator()
    threads = tf.train.start_queue_runners(sess=s, coord=coord)

    try:
        step = 0
        epoch = 0
        batch_within_epoch = 0
        while not coord.should_stop() and epoch < N_EPOCHS:
            xa, xb = s.run(batch)
            step += 1
            print(str(step), 'aa=', str(xa), 'bb=', str(xb))
            batch_within_epoch += 1
            if batch_within_epoch % steps_per_epoch == 0:
                print('------')
                epoch += 1
                batch_within_epoch = 0
    except tf.errors.OutOfRangeError:
        print('out of range')
    finally:
        coord.request_stop()
        print('requesting stop')
```

Works fine if the `num_epochs` parameter to `slice_input_producer` is removed.

**Comment by fwalch** [comment]
@rgobbel I think you just need to move the creation of the init ops below the creation of `slice_input_producer` to fix that.

**Comment by rgobbel** [comment]
Thanks! That fixed it. Turns out my real model has other issues, based on my obviously incorrect understanding of how queues and batches work.

**Comment by krishpop** [comment]
```
with tf.Graph().as_default():
    train_data = tf.constant(self.train.values)
    batched_data = tf.train.slice_input_producer([train_data], num_epochs=self.config.num_epochs)
    batch = tf.train.batch([batched_data], batch_size=self.config.batch_size)

    batch_pl = tf.placeholder(tf.float32, shape=(self.config.batch_size, self.train.shape[1]))
    encoded, decoded = self.inference(batch_pl)
    loss = self.loss(batch_pl, encoded, decoded)
    train_op, global_step = self.training(loss)

    # Build the summary Tensor based on the TF collection of Summaries.
    summary_op = tf.summary.merge_all()
    # Add the variable initializer Op.
    init = tf.group(tf.global_variables_initializer(),
                    tf.local_variables_initializer())
    # Create a saver for writing training checkpoints.
    saver = tf.train.Saver(write_version=tf.train.SaverDef.V2)
    # Create a session for running Ops on the Graph.
    sess = tf.Session()
    # Instantiate a SummaryWriter to output summaries and the Graph.
    summary_writer = tf.summary.FileWriter(ckpt_dir, sess.graph)

    # Start input enqueue threads.
    coord = tf.train.Coordinator()
    threads = tf.train.start_queue_runners(sess=sess, coord=coord)

    sess.run(init)
